File: Catalog REST/src/main/java/cz/vhromada/catalog/rest/controller/ProgramController.java
package cz.vhromada.catalog.rest.controller;

import cz.vhromada.catalog.facade.ProgramFacade;
import cz.vhromada.catalog.facade.to.ProgramTO;
import cz.vhromada.validators.exceptions.RecordNotFoundException;
import cz.vhromada.validators.exceptions.ValidationException;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.ResponseBody;

/**
 * A class represents controller for programs.
 *
 * @author Vladimir Hromada
 */
@Controller("programController")
@RequestMapping("/programs")
public class ProgramController extends JsonController {

	@Autowired
	@Qualifier("programFacade")
	private ProgramFacade programFacade;

	/** Creates new data. */
	@RequestMapping(value = "/new", method = RequestMethod.GET)
	@ResponseBody
	public void newData() {
		programFacade.newData();
	}

	/**
	 * Returns list of programs.
	 *
	 * @return list of programs
	 */
	@RequestMapping(value = { "", "/", "list" }, method = RequestMethod.GET)
	@ResponseBody
	public String getPrograms() {
		return serialize(programFacade.getPrograms());
	}

	/**
	 * Returns program with ID or null if there isn't such program.
	 *
	 * @param id ID
	 * @return program with ID or null if there isn't such program
	 * @throws IllegalArgumentException if ID is null
	 */
	@RequestMapping(value = "/{id}", method = RequestMethod.GET)
	@ResponseBody
	public String getProgram(@PathVariable("id") final Integer id) {
		return serialize(programFacade.getProgram(id));
	}

	/**
	 * Adds program. Sets new ID and position.
	 *
	 * @param program program
	 * @throws IllegalArgumentException if program is null
	 * @throws ValidationException      if ID isn't null
	 *                                  or name is null
	 *                                  or name is empty string
	 *                                  or URL to english Wikipedia page about program is null
	 *                                  or URL to czech Wikipedia page about program is null
	 *                                  or count of media isn't positive number
	 *                                  or other data is null
	 *                                  or note is null
	 */
	@RequestMapping(value = "/add", method = RequestMethod.POST)
	@ResponseBody
	public void add(final String program) {
		programFacade.add(deserialize(program, ProgramTO.class));
	}

	/**
	 * Updates program.
	 *
	 * @param program new value of program
	 * @throws IllegalArgumentException if program is null
	 * @throws ValidationException      if ID is null
	 *                                  or name is null
	 *                                  or name is empty string
	 *                                  or URL to english Wikipedia page about program is null
	 *                                  or URL to czech Wikipedia page about program is null
	 *                                  or count of media isn't positive number
	 *                                  or other data is null
	 *                                  or note is null
	 * @throws RecordNotFoundException  if program doesn't exist in data storage
	 */
	@RequestMapping(value = "/update", method = RequestMethod.POST)
	@ResponseBody
	public void update(final String program) {
		programFacade.update(deserialize(program, ProgramTO.class));
	}

	/**
	 * Removes program.
	 *
	 * @param program program
	 * @throws IllegalArgumentException if program is null
	 * @throws ValidationException      if ID is null
	 * @throws RecordNotFoundException  if program doesn't exist in data storage
	 */
	@RequestMapping(value = "/remove", method = RequestMethod.POST)
	@ResponseBody
	public void remove(final String program) {
		programFacade.remove(deserialize(program, ProgramTO.class));
	}

	/**
	 * Duplicates program.
	 *
	 * @param program program
	 * @throws IllegalArgumentException if program is null
	 * @throws ValidationException      if ID is null
	 * @throws RecordNotFoundException  if program doesn't exist in data storage
	 */
	@RequestMapping(value = "/duplicate", method = RequestMethod.POST)
	@ResponseBody
	public void duplicate(final String program) {
		programFacade.duplicate(deserialize(program, ProgramTO.class));
	}

	/**
	 * Moves program in list one position up.
	 *
	 * @param program program
	 * @throws IllegalArgumentException if program is null
	 * @throws ValidationException      if ID is null
	 *                                  or program can't be moved up
	 * @throws RecordNotFoundException  if program doesn't exist in data storage
	 */
	@RequestMapping(value = "/moveUp", method = RequestMethod.POST)
	@ResponseBody
	public void moveUp(final String program) {
		programFacade.moveUp(deserialize(program, ProgramTO.class));
	}

	/**
	 * Moves program in list one position down.
	 *
	 * @param program program
	 * @throws IllegalArgumentException if program is null
	 * @throws ValidationException      if ID is null
	 *                                  or program can't be moved down
	 * @throws RecordNotFoundException  if program doesn't exist in data storage
	 */
	@RequestMapping(value = "/moveDown", method = RequestMethod.POST)
	@ResponseBody
	public void moveDown(final String program) {
		programFacade.moveDown(deserialize(program, ProgramTO.class));
	}

	/**
	 * Returns true if program exists.
	 *
	 * @param program program
	 * @return true if program exists
	 * @throws IllegalArgumentException if program is null
	 * @throws ValidationException      if ID is null
	 */
	@RequestMapping(value = "/exists", method = RequestMethod.POST)
	@ResponseBody
	public String exists(final String program) {
		return serialize(programFacade.exists(deserialize(program, ProgramTO.class)));
	}

	/** Updates positions. */
	@RequestMapping(value = "/updatePositions", method = RequestMethod.GET)
	@ResponseBody
	public void updatePositions() {
		programFacade.updatePositions();
	}

	/**
	 * Returns total count of media.
	 *
	 * @return total count of media
	 */
	@RequestMapping(value = "/totalMedia", method = RequestMethod.GET)
	@ResponseBody
	public String getTotalMediaCount() {
		return serialize(programFacade.getTotalMediaCount());
	}

}

File: catalog-rest/src/main/kotlin/com/github/vhromada/catalog/rest/controller/ProgramController.kt
package com.github.vhromada.catalog.rest.controller

import com.github.vhromada.catalog.entity.Program
import com.github.vhromada.catalog.facade.ProgramFacade
import com.github.vhromada.common.web.controller.AbstractController
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestController

/**
 * A class represents controller for programs.
 *
 * @author Vladimir Hromada
 */
@RestController("programController")
@RequestMapping("/catalog/programs")
class ProgramController(
    private val programFacade: ProgramFacade
) : AbstractController() {

    /**
     * Returns list of programs.
     *
     * @return list of programs
     */
    @GetMapping
    fun getPrograms(): List<Program> {
        return processResult(result = programFacade.getAll())!!
    }

    /**
     * Creates new data.
     */
    @PostMapping("/new")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun newData() {
        processResult(result = programFacade.newData())
    }

    /**
     * Returns program with ID.
     *
     * @param id ID
     * @return program with ID
     */
    @GetMapping("/{id}")
    fun getProgram(@PathVariable("id") id: Int): Program {
        return processResult(result = programFacade.get(id = id))!!
    }

    /**
     * Adds program. Sets new ID and position.
     * <br></br>
     * Validation errors:
     *
     *  * ID isn't null
     *  * Position isn't null
     *  * Name is null
     *  * Name is empty string
     *  * URL to english Wikipedia page about program is null
     *  * URL to czech Wikipedia page about program is null
     *  * Count of media isn't positive number
     *  * Format is null
     *  * Other data is null
     *  * Note is null
     *
     * @param program program
     */
    @PutMapping("/add")
    @ResponseStatus(HttpStatus.CREATED)
    fun add(@RequestBody program: Program) {
        processResult(result = programFacade.add(data = program))
    }

    /**
     * Updates program.
     * <br></br>
     * Validation errors:
     *
     *  * ID is null
     *  * Position is null
     *  * Name is null
     *  * Name is empty string
     *  * URL to english Wikipedia page about program is null
     *  * URL to czech Wikipedia page about program is null
     *  * Count of media isn't positive number
     *  * Format is null
     *  * Other data is null
     *  * Note is null
     *  * Program doesn't exist in data storage
     *
     * @param program new value of program
     */
    @PostMapping("/update")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun update(@RequestBody program: Program) {
        processResult(result = programFacade.update(data = program))
    }

    /**
     * Removes program.
     * <br></br>
     * Validation errors:
     *
     *  * Program doesn't exist in data storage
     *
     * @param id ID
     */
    @DeleteMapping("/remove/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun remove(@PathVariable("id") id: Int) {
        processResult(result = programFacade.remove(id = id))
    }

    /**
     * Duplicates program.
     * <br></br>
     * Validation errors:
     *
     *  * Program doesn't exist in data storage
     *
     * @param id ID
     */
    @PostMapping("/duplicate/{id}")
    @ResponseStatus(HttpStatus.CREATED)
    fun duplicate(@PathVariable("id") id: Int) {
        processResult(result = programFacade.duplicate(id = id))
    }

    /**
     * Moves program in list one position up.
     * <br></br>
     * Validation errors:
     *
     *  * Program can't be moved up
     *  * Program doesn't exist in data storage
     *
     * @param id ID
     */
    @PostMapping("/moveUp/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun moveUp(@PathVariable("id") id: Int) {
        processResult(result = programFacade.moveUp(id = id))
    }

    /**
     * Moves program in list one position down.
     * <br></br>
     * Validation errors:
     *
     *  * Program can't be moved down
     *  * Program doesn't exist in data storage
     *
     * @param id ID
     */
    @PostMapping("/moveDown/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun moveDown(@PathVariable("id") id: Int) {
        processResult(result = programFacade.moveDown(id = id))
    }

    /**
     * Updates positions.
     */
    @PostMapping("/updatePositions")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun updatePositions() {
        processResult(result = programFacade.updatePositions())
    }

    /**
     * Returns total count of media.
     *
     * @return total count of media
     */
    @GetMapping("/totalMedia")
    fun totalMediaCount(): Int {
        return processResult(result = programFacade.getTotalMediaCount())!!
    }

}
